daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -943,32 +943,31 @@ private[lf] final class Compiler(
   ) =
     let(SBUFetch(tmplId)(svar(cidPos))) { tmplArgPos =>
       addExprVar(tmpl.param, tmplArgPos)
-      let(
-        SBUBeginExercise(tmplId, choice.name, choice.consuming, byKey = mbKey.isDefined)(
-          svar(choiceArgPos),
-          svar(cidPos),
-          compile(tmpl.signatories),
-          compile(tmpl.observers), //
-          {
-            addExprVar(choice.argBinder._1, choiceArgPos)
-            compile(choice.controllers)
-          }, //
-          {
-            choice.choiceObservers match {
-              case Some(observers) => compile(observers)
-              case None => SEValue.EmptyList
-            }
-          },
-          mbKey.fold(compileKeyWithMaintainers(tmpl.key))(pos => SBSome(svar(pos))),
-        )
-      ) { _ =>
-        addExprVar(choice.selfBinder, cidPos)
-        let(app(compile(choice.update), svar(tokenPos))) { retValuePos =>
-          let(SBUEndExercise(tmplId)(svar(retValuePos))) { _ =>
-            svar(retValuePos)
-          }
-        }
-      }
+      SEScopeExercise(
```

Comment:
Why do we need a separate `SEScopeExercise` can we just push the continuation in BeginExercise or alternatively push what we do in beginexercise in scopexercise?

Reply:
You can't push the continuation in `SBUBeginExercise` because the scope is compiled to encompass the _begin_ and the _update_.

I considered moving/inlining the code of the existing `SBUBeginExercise` into the new `SEScopeExercise`, but this seemed quite disruptive (making the PR harder to review) with minimal benefit if any.

Reply:
Makes sense :+1: